### PR TITLE
MM-47095 Migrate AbstractIncomingWebhook to TypeScript

### LINKED ISSUE #21074: Migrate "components/integrations/abstract_incoming_webhook.jsx" to Typescript
We are plan to migrate the mattermost-webapp over to TypeScript to facilitate better code quality. This `Help Wanted` issue is to modify the above mentioned file and associate test files (if the file exists).

Below is a checklist of items you should be doing to migrate:

  * Rename the files to their associated TypeScript extensions (ie. `js` to `ts`, `jsx` to `tsx`).
  * Add relevant types in the component(s) code.
  * Update any components using those to the correct imports.
  * Fix any errors generated by the TypeScript compiler. You can run `rpm run check-types` to display any errors.

We are aiming to be as strict as possible for the time being, but if you run into any issues where the file cannot be successfully migrated without significant change in other libraries, or our `tsconfig.json`, please feel free to raise them as we are aiming to keep this process smooth and gradual, while enforcing as much as we can.

Some examples of already migrated files can be found here:

  * [https://github.com/mattermost/mattermost-webapp/pull/10826](https://github.com/mattermost/mattermost-webapp/pull/10826|smart-link) 
  * [https://github.com/mattermost/mattermost-webapp/pull/10931](https://github.com/mattermost/mattermost-webapp/pull/10931|smart-link) 

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-47095

### PR DESCRIPTION
**Pull request by Dismated**
#### Summary
   Migrates file from JS to TS

#### Ticket Link
   Fixes https://github.com/mattermost/mattermost-server/issues/21074
   https://mattermost.atlassian.net/browse/MM-47095

### CODE REVIEW

File: components/advanced_create_comment/advanced_create_comment.test.tsx [hunk 1]
```diff
@@ -0,0 +1,1773 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (hmhealey) [inline components/advanced_create_comment/advanced_create_comment.test.tsx:1]: Did you mean to submit this file with this PR? It seems unrelated to the ticket

File: components/integrations/abstract_incoming_webhook.tsx [hunk 2]
```diff
@@ -10,64 +9,90 @@ import BackstageHeader from 'components/backstage/components/backstage_header';
...
-        * Whether to allow configuration of the default post icon.
-        */
-        enablePostIconOverride: PropTypes.bool.isRequired,
-
-        /**
-        * The async function to run when the action button is pressed
-        */
-        action: PropTypes.func.isRequired,
-    }
+interface Hook {
```
Comment (hmhealey) [inline components/integrations/abstract_incoming_webhook.tsx:15]: Instead of adding a new type for this, could we use `IncomingWebhook` from `@mattermost/types/integrations` for this or `Partial<IncomingWebhook>`?

File: components/integrations/abstract_incoming_webhook.tsx [hunk 3]
```diff
@@ -10,64 +9,90 @@ import BackstageHeader from 'components/backstage/components/backstage_header';
...
+
+    /**
+    * The current team
+    */
+    team: Team;
+
+    /**
+    * The header text to render, has id and defaultMessage
+    */
+    header: {id: string; defaultMessage: string};
```
Comment (hmhealey) [inline components/integrations/abstract_incoming_webhook.tsx:46]: `header`, `footer`, and `loading` could use the `MessageDescriptor` type from `react-intl`

File: components/advanced_create_comment/advanced_create_comment.test.tsx (hunk 1 above)
Comment (hmhealey) [reply components/advanced_create_comment/advanced_create_comment.test.tsx:1]: It's also what's causing the lint and type-check CI jobs to fail as well

**Review by hmhealey — CHANGES_REQUESTED** [review]
Thanks for the PR! This looks pretty good, but I've got a couple suggestions for where we can reuse some existing types

File: components/advanced_create_comment/advanced_create_comment.test.tsx (hunk 1 above)
Comment (Dismated) [reply components/advanced_create_comment/advanced_create_comment.test.tsx:1]: My bad, somehow I messed up my branches and didn't see the file coming through commit. Fixing it right now.

**Review by ghost — APPROVED** [review]

File: components/advanced_create_comment/advanced_create_comment.test.tsx (hunk 1 above)
Comment (hmhealey) [reply components/advanced_create_comment/advanced_create_comment.test.tsx]: No worries!

**Review by hmhealey — APPROVED** [review]
Thanks for changing those types! This looks good to me now

**Review by saturninoabril — COMMENTED** [review]
There's an issue with CI build. I already informed the Release team. I'll get back on this once resolved.

File: components/integrations/add_incoming_webhook/add_incoming_webhook.tsx [hunk 4]
```diff
@@ -41,6 +41,7 @@ type Props = {
 
 type State = {
     serverError: string;
+
```
Comment (saturninoabril) [inline components/integrations/add_incoming_webhook/add_incoming_webhook.tsx:44]: Please revert unnecessary change here.
```suggestion
```

File: components/integrations/abstract_incoming_webhook.tsx [hunk 5]
```diff
@@ -1,87 +1,100 @@
...
+    serverError: string;
 
-        /**
-        * Whether to allow configuration of the default post username.
-        */
-        enablePostUsernameOverride: PropTypes.bool.isRequired,
+    /**
+    * The hook used to set the initial state
+    */
+    initialHook?: IncomingWebhook | Record<string, never>;
```
Comment (M-ZubairAhmed) [inline components/integrations/abstract_incoming_webhook.tsx:58]: why do we use never here?

File: components/integrations/abstract_incoming_webhook.tsx [hunk 6]
```diff
@@ -115,50 +128,56 @@ export default class AbstractIncomingWebhook extends React.PureComponent {
...
+            delete_at: this.props.initialHook?.delete_at || 0,
+            team_id: this.props.initialHook?.team_id || '',
+            user_id: this.props.initialHook?.user_id || '',
         };
 
         this.props.action(hook).then(() => this.setState({saving: false}));
     }
 
-    updateDisplayName = (e) => {
+    updateDisplayName: React.ChangeEventHandler<HTMLInputElement> = (e) => {
```
Comment (M-ZubairAhmed) [inline components/integrations/abstract_incoming_webhook.tsx:142]: please selectively import directly from react
```import {ChangeEventHandler} from 'react'```
In other places also

File: components/integrations/abstract_incoming_webhook.tsx (hunk 5 above)
Comment (Dismated) [reply components/integrations/abstract_incoming_webhook.tsx:58]: Because the initialHook value can be {}. Record<string, never> is an empty object in the sense that you can't add any string-type accessor properties to it, so it will always be empty.

**Comment by M-ZubairAhmed** [comment]
Hey there!, Thank you for your PR.
Webapp is now part of mattermost monorepo now at https://github.com/mattermost/mattermost-server/tree/master/webapp Would you be interested in opening the PR over there?

**Comment by hmhealey** [comment]
The web app now lives in the `webapp/channels` over in https://github.com/mattermost/mattermost-server. We decided to move it to be a monorepo to make both development and our CI pipelines easier, but it's unfortunately disruptive right now. We've updated [our setup docs](https://developers.mattermost.com/contribute/developer-setup/) and added [some steps for migrating to working with the monorepo](https://developers.mattermost.com/contribute/monorepo-migration-notes/).

Copying the changes over there hopefully shouldn't be too hard. There's a few changes to file paths (such as the code living in a dedicated `src` subdirectory), but most things should look the same as they do now. 

Let us know if you could use any help with this. Hopefully the process is smooth.
